README.md
## PyOTP - Python One Time Password Library

A Python library for generating one time passcodes according to [RFC
6238](http://tools.ietf.org/html/rfc6238). This version is ported from
https://github.com/nathforge/pyotp to generate accurate 6 digit codes.

This is compatible with Google Authenticator apps available for Android and
iPhone, and now in use on GMail

### Quick overview of using One Time Passwords on your phone

* OTP's involve a shared secret, stored both on the phone and the server
* OTP's can be generated on a phone without internet connectivity(AT&T mode)
* OTP's should always be used as a second factor of authentication(if your
    phone is lost, you account is still secured with a password) * Google
    Authenticator allows you to store multiple OTP secrets and provision those
    using a QR Code(no more typing in the secret)

### Usage

#### Time based OTP's

    totp = pyotp.TOTP('base32secret3232') totp.now() # => 492039

    # OTP verified for current time totp.verify(492039) # => True
    time.sleep(30) totp.verify(492039) # => False

#### Counter based OTP's

    hotp = pyotp.HOTP('base32secret3232') hotp.at(0) # => 260182 hotp.at(1)
    # => 55283 hotp.at(1401) # => 316439

    # OTP verified with a counter hotp.verify(316439, 1401) # => True
    hotp.verify(316439, 1402) # => False

#### Generating a Base32 Secret key

    pyotp.random_base32()

#### Google Authenticator Compatible

The library works with the Google Authenticator iPhone and Android app, and
also includes the ability to generate provisioning URI's for use with the QR
Code scanner built into the app.

    totp.provisioning_uri("alice@google.com") # =>
    'otpauth://totp/alice@google.com?secret=JBSWY3DPEHPK3PXP'

    hotp.provisioning_uri("alice@google.com", 0) # =>
    'otpauth://hotp/alice@google.com?secret=JBSWY3DPEHPK3PXP&counter=0'

This can then be rendered as a QR Code which can then be scanned and added to
the users list of OTP credentials.

##### Working example

Scan the following barcode with your phone, using Google Authenticator

![QR Code for
OTP](http://chart.apis.google.com/chart?cht=qr&chs=250x250&chl=otpauth%3A%2F%2Ftotp%2Falice%40google.com%3Fsecret%3DJBSWY3DPEHPK3PXP)

Now run the following and compare the output

    import pyotp totp = pyotp.TOTP("JBSWY3DPEHPK3PXP")
    print "Current OTP: %s" % totp.now()

### See also:

Original Ruby version of ROTP by [Mark Percival](https://github.com/mdp)
- [ROTP](https://github.com/mdp/rotp) PHP port of ROTP by [Le
    Lag](https://github.com/lelag) - [OTPHP](https://github.com/lelag/otphp)


Changes.rst
Changes for v2.5.0 (2021-01-29)
===============================

-  Add optional image parameter to provisioning_uri (#113)

-  Support for 7-digit codes in ‘parse_uri’ (#111)

-  Raise default and minimum base32 secret length to 26

Changes for v2.4.1 (2020-10-16)
===============================

-  parse_uri: Fix handling of period, counter (#108)

-  Add support for timezone aware datetime as argument to
   ``TOTP.timecode()`` (#107)

Changes for v2.4.0 (2020-07-29)
===============================

-  Fix data type for at(for_time) (#85)

-  Add support for parsing provisioning URIs (#84)

-  Raise error when trying to generate secret that is too short (The
   secret must be at least 128 bits)

-  Add random_hex function (#82)

Changes for v2.3.0 (2019-07-26)
===============================

-  Fix comparison behavior on Python 2.7

Changes for v2.2.8 (2019-07-26)
===============================

-  Fix comparison of unicode chars (#78)

-  Minor documentation and test fixes

Changes for v2.2.7 (2018-11-05)
===============================

-  Have random_base32() use ‘secrets’ as rand source (#66)

-  Documentation: Add security considerations, minimal security
   checklist, other improvements

-  Update setup.py to reference correct license

Changes for v2.2.6 (2017-06-10)
===============================

-  Fix tests wrt double-quoting in provisioning URIs

Changes for v2.2.5 (2017-06-03)
===============================

-  Quote issuer QS parameter in provisioning\_uri. Fixes #47.

-  Raise an exception if a negative integer is passed to at() (#41).

-  Documentation and release infrastructure improvements.

Changes for v2.2.4 (2017-01-04)
===============================

-  Restore Python 2.6 compatibility (however, Python 2.6 is not
   supported)

-  Documentation and test improvements

-  Fix release infra script, part 2

Changes for v2.2.3 (2017-01-04)
===============================

-  Restore Python 2.6 compatibility (however, Python 2.6 is not
   supported)

-  Documentation and test improvements

-  Fix release infra script

Changes for v2.2.2 (2017-01-04)
===============================

-  Restore Python 2.6 compatibility (however, Python 2.6 is not
   supported)

-  Documentation and test improvements

Changes for v2.2.1 (2016-08-30)
===============================

-  Avoid using python-future; it has subdependencies that limit
   compatibility (#34)
-  Make test suite pass on 32-bit platforms (#30)
-  Timing attack resistance fix: don't reveal string length to attacker.
   Thanks to Eeo Jun (#28).
-  Support algorithm, digits, period parameters in provisioning\_uri.
   Thanks to Dionisio E Alonso (#33).
-  Minor style and packaging infrastructure fixes.

Changes for v2.2.0 (2016-08-30)
===============================

-  See v2.2.1

Version 2.1.0 (2016-05-02)
--------------------------
- Add extended range support to TOTP.verify. Thanks to Zeev Rotshtein (PR #19).
- Handle missing padding of encoded secret. Thanks to Kun Yan (#20).
- Miscellaneous fixes.

Version 2.0.1 (2015-09-28)
--------------------------
- Fix packaging issue in v2.0.0 that prevented installation with easy_install.

Version 2.0.0 (2015-08-22)
--------------------------
- The ``pyotp.HOTP.at()``, ``pyotp.TOTP.at()``, and
  ``pyotp.TOTP.now()`` methods now return strings instead of
  integers. Thanks to Rohan Dhaimade (PR #16).

Version 1.4.2 (2015-07-21)
--------------------------
- Begin tracking changes in change log.
- Update documentation.
- Introduce Travis CI integration.

Version 1.3.1 (2012-02-29)
--------------------------
- Initial release.


src/pyotp/__init__.py
import hashlib
from re import split
from typing import Any, Dict, Sequence
from urllib.parse import unquote, urlparse, parse_qsl

from .compat import random
from .hotp import HOTP as HOTP
from .otp import OTP as OTP
from .totp import TOTP as TOTP


def random_base32(length: int = 26, chars: Sequence[str] = list('ABCDEFGHIJKLMNOPQRSTUVWXYZ234567')) -> str:
    if length < 26:
        raise ValueError("Secrets should be at least 128 bits")

    return ''.join(
        random.choice(chars)
        for _ in range(length)
    )


def random_hex(length: int = 32, chars: Sequence[str] = list('ABCDEF0123456789')) -> str:
    if length < 32:
        raise ValueError("Secrets should be at least 128 bits")
    return random_base32(length=length, chars=chars)


def parse_uri(uri: str) -> OTP:
    """
    Parses the provisioning URI for the OTP; works for either TOTP or HOTP.

    See also:
        https://github.com/google/google-authenticator/wiki/Key-Uri-Format

    :param uri: the hotp/totp URI to parse
    :returns: OTP object
    """

    # Secret (to be filled in later)
    secret = None

    # Data we'll parse to the correct constructor
    otp_data = {}  # type: Dict[str, Any]

    # Parse with URLlib
    parsed_uri = urlparse(unquote(uri))

    if parsed_uri.scheme != 'otpauth':
        raise ValueError('Not an otpauth URI')

    # Parse issuer/accountname info
    accountinfo_parts = split(':|%3A', parsed_uri.path[1:], maxsplit=1)
    if len(accountinfo_parts) == 1:
        otp_data['name'] = accountinfo_parts[0]
    else:
        otp_data['issuer'] = accountinfo_parts[0]
        otp_data['name'] = accountinfo_parts[1]

    # Parse values
    for key, value in parse_qsl(parsed_uri.query):
        if key == 'secret':
            secret = value
        elif key == 'issuer':
            if 'issuer' in otp_data and otp_data['issuer'] is not None and otp_data['issuer'] != value:
                raise ValueError('If issuer is specified in both label and parameters, it should be equal.')
            otp_data['issuer'] = value
        elif key == 'algorithm':
            if value == 'SHA1':
                otp_data['digest'] = hashlib.sha1
            elif value == 'SHA256':
                otp_data['digest'] = hashlib.sha256
            elif value == 'SHA512':
                otp_data['digest'] = hashlib.sha512
            else:
                raise ValueError('Invalid value for algorithm, must be SHA1, SHA256 or SHA512')
        elif key == 'digits':
            digits = int(value)
            if digits not in [6, 7, 8]:
                raise ValueError('Digits may only be 6, 7, or 8')
            otp_data['digits'] = digits
        elif key == 'period':
            otp_data['interval'] = int(value)
        elif key == 'counter':
            otp_data['initial_count'] = int(value)
        elif key != 'image':
            raise ValueError('{} is not a valid parameter'.format(key))

    if not secret:
        raise ValueError('No secret found in URI')

    # Create objects
    if parsed_uri.netloc == 'totp':
        return TOTP(secret, **otp_data)
    elif parsed_uri.netloc == 'hotp':
        return HOTP(secret, **otp_data)

    raise ValueError('Not a supported OTP type')
